### ISSUE TITLE
Minor mode names cut without mode-line-end-spaces="   "

### ISSUE THREAD

**Issue by pronobis** [Malabarba/smart-mode-line#157]
I'm having the following problem with minor mode names being cut:
![modeline](https://cloud.githubusercontent.com/assets/331795/11017346/03fdb5a4-8551-11e5-91da-ca9ba8f5a9bc.png)

Setting `mode-line-end-spaces` to some value solves the problem but SML resets it in it's setup for some reason I assume.

**Comment by Malabarba** [comment]
If the value of this variable is being changed, it's probably because it's been accidentally saved by custom-set-variables. Check your custom settings (these are the settings inside a call to custom-set-variables), and see if you find this variable in there.

**Comment by pronobis** [comment]
I don't have it there. Anyway, that's not the main issue. Using that value is just a workaround. Do you know why sml could be cutting the minor modes?

**Comment by Malabarba** [comment]
Sorry for the late response. 

> Do you know why sml could be cutting the minor modes?

Either the window width is not being properly calculated, or you have some characters in your mode line that are not fixed-width. Could it be the latter?

**Comment by pronobis** [comment]
The only non-ascii characters I have there are the the arrows, which might potentially be wider than other characters. Is fixed width of each character assumed? 

Also, I'm using zoom-frm.el and when zooming in, the minor modes disappear outside the screen. Is there some refreshing that needs to happen after the change of font size?

**Comment by Malabarba** [comment]
> The only non-ascii characters I have there are the the arrows, which might potentially be wider than other characters. Is fixed width of each character assumed?

Yes, that's probably it. Emacs doesn't have very good ways of calculating the pixel-width of arbitrary strings.
You can try setting `sml/mode-width` to a fixed number. Or you can set it to `0` and set `sml/shorten-modes` to `nil` (which gives the stock mode-line behavior).

> Also, I'm using zoom-frm.el and when zooming in, the minor modes disappear outside the screen. Is there some refreshing that needs to happen after the change of font size?

You can try refreshing the mode-line with `force-modeline-update`, but it's unlikely to help here. I'm not familiar with zoom-frm, but I would guess it increases the font-size by using local faces, which means sml can't do anything about them either.
You'll probably have more luck just increasing the size of the `default` face, which sml should automatically adapt to.

    (set-face-attribute 'default nil :height 160)

**Comment by pronobis** [comment]
I removed the arrows and and the problem remains. It seems that the calculation of the empty space when `sml/mode-width` is set to `right` is somehow wrong pushing the minor modes outside the screen.

**Comment by Malabarba** [comment]
Could you try this again on the latest snapshot (should take a couple of hours to reach Melpa)?
#164 might have affected this.

**Comment by pronobis** [comment]
That indeed solved the problem of required spaces at the end. Thanks! The minor modes still end up outside the screen when zooming in with zoom-frm.el though.

**Comment by Malabarba** [comment]
Yes, that last part is unlikely to change until I have a LOT of free time. :-(

**Comment by lyro** [comment]
I've played a little bit around with zoom-frm and observed the following: when zooming, the value returned by `(window-total-width)` does not change. However, smart-mode-line depends on that value to compute the width of the modeline, and, in fact, that value **should** change in my opinion. But if you split a window **vertically** (and maybe close the new window again), then that value is updated and then the modeline is displayed correctly. Does this work for you, too?

**Comment by Malabarba** [comment]
Hm. Yes, sml relies on window-total-width, so that's probably the source of
the problem. I agree with you that its return value should change when you
zoom in and out, so maybe this is a zoom-frm bug.

(I can't test if splitting the window works for me right now.)

**Comment by pronobis** [comment]
Seems that this is a bug in Emacs: https://lists.gnu.org/archive/html/bug-gnu-emacs/2015-03/msg00003.html
